gensim/models/hdpmodel.py

```diff
@@ -339,6 +339,8 @@ def __init__(self, corpus, id2word, max_chunks=None, max_time=None,
             corresponding lda model from :meth:`~gensim.models.hdpmodel.HdpModel.suggested_lda_model`
 
         """
+        if len(corpus) == 0:
```

Comment:
Is `corpus` guaranteed to support `len()` in HdpModel? What if it's just a generator?

Reply:
Most definitely not. The `len` exception is still there, plus now you're serializing into entire corpus into RAM! If it's a generator, it will be gone.

Does HdpModel support generators? Or are we guaranteed to have a repeatable iterable with `len`?

Reply:
ah ok got it, well the comments in the file just say 
```
corpus : iterable of list of (int, float)
            Corpus in BoW format.
```
so I wouldn't call this a guarantee but it does sound like list is the intended type. Should there maybe be type checking to guarantee that list is used?

Reply:
It's not a list. It's an _iterable_ of something.

That something is a list of tuples, but that still doesn't change the fact that you're dealing with an iterable, and you won't necessarily be able to call `len()` on it.

Reply:
oh I see, I understand now. it looks like `len()` is already being called on the corpus in another place in the constructor
```
if corpus:
        self.m_D = len(corpus)
```
leading me to believe that the corpus does in fact need to be a list..or should this be changed as well?

Reply:
would it maybe be easiest to use the existing utils.is_corpus method in the constructor? since it covers the empty corpus case and works for generators

Reply:
Yeah, I'd say calling `len(corpus)` is probably a bug, because it risks consuming the iterable.

Reply:
It's possible the HDP model does not support iterables. It is possible it supports `list` only :( (unlike other models in Gensim)

That's why I ask about the contract of HdpModel – whether it explicitly needs its input data in RAM (~list), or if it supports streamed inputs (~any iterable).

Reply:
@piskvorky I'll look over the code and also look around for usage of the model with a non-list iterable. if it turns out that only list is supported, would it be helpful to add support for all iterables? or not worth the effort since the model may be deprecated soon

Reply:
> I'll look over the code and also look around for usage of the model with a non-list iterable.

That will be great, thanks.

> or not worth the effort since the model may be deprecated soon

I'd say "not worth the effort" right now. It isprobably possible to clean up and optimize HDP, but we lack a clear motivating use-case. So, before any technical optimizations, we'd need a **good strong tutorial / how-to document**, showing off *why* HDP is worth it, how it's better than other models, when to use it.

The performance considerations are secondary to my mind.